crates/dojo/types/src/primitive.rs

```diff
@@ -403,6 +403,42 @@ impl Primitive {
             }
         }
     }
+
+    pub fn can_upgrade_to(&self, other: &Primitive) -> bool {
+        use Primitive::*;
+        match (self, other) {
+            // Boolean (no upgrades)
+            (Bool(_), Bool(_)) => true,
+
+            // Unsigned integer upgrades
+            (U8(_), U16(_) | U32(_) | USize(_) | U64(_) | U128(_) | Felt252(_)) => true,
+            (U16(_), U32(_) | USize(_) | U64(_) | U128(_) | Felt252(_)) => true,
+            (U32(_), USize(_) | U64(_) | U128(_) | Felt252(_)) => true,
+            (USize(_), U32(_) | U64(_) | U128(_) | Felt252(_)) => true,
```

Comment:
I've left a comment here https://github.com/dojoengine/dojo/pull/2925#discussion_r1929474559, I think having `usize` non-changeable would make more sense at the cairo level (for Torii, it's way less relevant, but it is to match the cairo rules).